### PR TITLE
Ability to mark flags as deprecated

### PR DESCRIPTION
**Pull request by eparis**
They will not show up in usage or help, but they will still work

### CODE REVIEW

**Comment by eparis** [comment]
@thockin So you have to define a flag, then call flagset.MarkDeprecated("name")

**Comment by eparis** [comment]
As an example in kubernetes I did

```
diff --git a/pkg/client/clientcmd/overrides.go b/pkg/client/clientcmd/overrides.go
index 160444b..a95fa43 100644
--- a/pkg/client/clientcmd/overrides.go
+++ b/pkg/client/clientcmd/overrides.go
@@ -134,6 +134,7 @@ func BindAuthInfoFlags(authInfo *clientcmdapi.AuthInfo, flags *pflag.FlagSet, fl
        flags.StringVar(&authInfo.ClientCertificate, flagNames.ClientCertificate, "", "Path to a client key file for TLS.")
        flags.StringVar(&authInfo.ClientKey, flagNames.ClientKey, "", "Path to a client key file for TLS.")
        flags.StringVar(&authInfo.Token, flagNames.Token, "", "Bearer token for authentication to the API server.")
+       flags.MarkDeprecated(flagNames.Token)
```

And the --token flag no longer shows up in kubectl help or docs

File: flag.go [hunk 1]
```diff
@@ -224,6 +225,16 @@ func (f *FlagSet) Lookup(name string) *Flag {
 	return f.formal[name]
 }
 
+// Mark a flag deprecated in your program
+func (f *FlagSet) MarkDeprecated(name string) error {
```
Comment (thockin) [inline flag.go]: same comment as subcommands - maybe there should be a string to help users figure out why it was deprecated, when it is going away, and what to use instead?

**Comment by eparis** [comment]
@thockin what do you think of this version?

File: flag.go [hunk 2]
```diff
@@ -243,6 +244,16 @@ func (f *FlagSet) lookup(name normalizedName) *Flag {
 	return f.formal[name]
 }
 
+// Mark a flag deprecated in your program
+func (f *FlagSet) MarkDeprecated(name string, usageMessage string) error {
+	flag := f.Lookup(name)
```
Comment (thockin) [inline flag.go]: You'll want to normalize the name now, I guess :)

**Comment by thockin** [comment]
love it, please test with normalized names :)

File: flag.go (hunk 2 above)
Comment (eparis) [inline flag.go]: I can't even count of the person writing code to know the names of their flags?   ok, i'll look at it.

Comment (eparis) [inline flag.go]: actually I think we're good, the f.Lookup() does the normalization.

```go
// Lookup returns the Flag structure of the named flag, returning nil if none exists.
func (f *FlagSet) Lookup(name string) *Flag {
        return f.lookup(f.normalizeFlagName(name))
}
```

**Comment by eparis** [comment]
@thockin Just added a new test which uses setWordSeparators()

File: flag.go (hunk 2 above)
Comment (thockin) [inline flag.go]: ah, right.  Or else it should compile-fail

**Comment by thockin** [comment]
LGTM

**Comment by bgrant0607** [comment]
Deprecated flags need to be fully functional for some time before we can cripple/ignore/remove them. This function doesn't permit type declarations nor default values.

**Comment by bgrant0607** [comment]
Nevermind.

**Pull request merged**